[cutlass backend] re-add pip cutlass path

Flamefire opened the pull request:
Revert #156651 to allow using the cutlass PIP package which is easier for users than the Git checkout or similar method.

Also fix a bug where the PIP cutlass path wouldn't be available to subprocesses spawned during benchmarking for algorithm selection. Looks like the "spawn" method does not inherit the (potentially) already set up `config.cuda.cutlass_dir` so in the subprocess the include paths will still be set to `"../third_party/cutlass/"` leading to compilation failure due to missing headers.

Ensure `try_import_cutlass` is called at that point, which due to caching is a no-op in most cases, so doesn't hurt.   
Change the logic to return `None` when cutlass isn't available returning more useful values for include paths, namely an empty list. This is in line with other inductor code which disables the CUTLASS backend when `try_import_cutlass` returns False

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

Flamefire committed:
Revert "[cutlass backend] delete pip cutlass path since nvidia stops supporting nvidia-cutlass (#156651)"

This reverts commit f97f03c7efcf2b7a45384b9094eb6be4cb419546.

Flamefire committed:
Ensure cutlass is initialized when include paths are determined

`try_import_cutlass` might set up a different CUTLASS include path
depending on the Python package installed
As this is used in other places where CUTLASS code is generated the same
must be used for include paths.

Otherwise compilation happening in subprocesses that might not have the
(modified) CUTLASS path set up will fail, e.g. the benchmarking codes
for algorithm selection.

Flamefire commented:
In a discussion with the maintainers of nvidia-cutlass they indeed deprecated the Python interface: https://github.com/NVIDIA/cutlass/discussions/2119

Question is for PyTorch: Is this still a supported backend?` If it requires manual fiddling with env variables and git checkouts it is not that user-friendly.   
Would it work with a "regular" installation of CUTLASS? That should install the headers too and usually makes them available to the compiler in some way (installation to default directory like `/usr`, using `$CPATH` or `$C_INCLUDE_PATH`)    
So what is the intended usage and installation method, especially for system administrators (we currently package PyTorch and dependencies in environment modules on HPC systems)?

This PR allows at least using the nvidia-cutlass 3.8.0.0 release even though there are newer versions of CUTLASS.

Skylion007 commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -89,6 +89,34 @@ def try_import_cutlass() -> bool:
+        return True
+    except ModuleNotFoundError:
+        log.debug(
+            "cutlass_library not found in sys.path, trying to import from config.cuda.cutlass_dir"
```
Add exc_info=True to provide additional details here about it not being able to find it

Skylion007 commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -89,6 +89,34 @@ def try_import_cutlass() -> bool:
+        import cutlass  # type: ignore[import-not-found]  # noqa: F811
+        import cutlass_library  # type: ignore[import-not-found]  # noqa: F811
+
+        cutlass_minor_vesion = int(cutlass.__version__.split(".")[1])
```
Um isn't the way of checking versions deprecated and we should use importlib metadata?

Flamefire replied:
Probably. I just reverted the commit that removed this restoring the original code and didn't want to change it.

Anyway as it's good going forward I did that, fixed the typo and actually check for ` < 3.7` and not fail for e.g. `4.1` when only checking the minor version.

Flamefire committed:
Use importlib.metadata, packaging.Version and log exception

Also fix comment.

henrylhtsang reviewed (approved):
Thanks a lot for helping!

Flamefire commented:
@pytorchbot rebase

atalman commented:
This broke nightly builds: https://github.com/pytorch/pytorch/actions/runs/16985419009/job/48153244921 - Will be reverting
```
    from ..codegen.cuda.gemm_template import CUTLASS2xGemmTemplate, CUTLASS3xGemmTemplate
  File "/Users/runner/work/_temp/anaconda/envs/test_conda_env/lib/python3.10/site-packages/torch/_inductor/codegen/cuda/gemm_template.py", line 14, in <module>
    from torch._inductor.codegen.cuda.cutlass_cache import maybe_fetch_ops
  File "/Users/runner/work/_temp/anaconda/envs/test_conda_env/lib/python3.10/site-packages/torch/_inductor/codegen/cuda/cutlass_cache.py", line 13, in <module>
    from torch._inductor.codegen.cuda import cutlass_utils, serialization
  File "/Users/runner/work/_temp/anaconda/envs/test_conda_env/lib/python3.10/site-packages/torch/_inductor/codegen/cuda/cutlass_utils.py", line 15, in <module>
    from packaging.version import Version
ModuleNotFoundError: No module named 'packaging'
```

atalman commented:
@pytorchmergebot revert -c nosignal -m "broke macos nightly"

Flamefire committed:
Use vendored packaging

Flamefire commented:
@pytorchbot rebase

etaf commented:
Hi @Flamefire, may we proceed to land this PR?
I’m working on enabling CUTLASS support for Intel GPUs, and since we rely on the pip installation flow for users, this PR is required to move forward.

Flamefire commented:
@etaf Conflicts caused by NVIDIA renaming the Python package from cutlass to cutlass_cppgen

Resolved by adding a fallback to cutlass if cutlass_cppgen can't be imported

Flamefire committed:
Use cutlass as fallback for cutlass_cppgen import

The package is renamed in 4.2.1 but allow for the older package name.

Flamefire committed:
Return cloned paths from `_clone_cutlass_paths`

The folders are explicitely copied to the build dir and hence should be used from there

etaf commented:
Hi，@mlazos  @henrylhtsang  would you kindly help to reveiw this PR?

Flamefire committed:
Fix package name when getting metadata

The package is always called "nvidia-cutlass", only the import name is
either `cutlass` or `cutlass_cppgen`.

Flamefire commented:
From my experiments it seems that cloning NVIDIA/CUTLASS using the same commit as referenced as a submodule by the used PyTorch version and setting `TORCHINDUCTOR_CUTLASS_DIR` to that works.

It doesn't look like it is easily possible to use PYPI packages anymore due to how PyTorch expects the package. Internally there seems to be a `python_cutlass` "package" that is custom-made similar to how PyTorch copies things from the CUTLASS repository at runtime with no clear correspondence to the official packages.

henrylhtsang commented:
> From my experiments it seems that cloning NVIDIA/CUTLASS using the same commit as referenced as a submodule by the used PyTorch version and setting `TORCHINDUCTOR_CUTLASS_DIR` to that works.
> 
> It doesn't look like it is easily possible to use PYPI packages anymore due to how PyTorch expects the package. Internally there seems to be a `python_cutlass` "package" that is custom-made similar to how PyTorch copies things from the CUTLASS repository at runtime with no clear correspondence to the official packages.

yeah maybe need to nudge nvidia to release 4.2.1, since they finishes the name change (cutlass -> cutlass_cppgen) in 4.2.1
